File: Sorting/Bubble sort/bubbleSort.js
function bubbleSort(arr){
    for(let i = 0; i < arr.length; i++){
        let isSwap = false  // if array is sorted then we break out loop
        for(let j = 0; j < arr.length - i - 1; j++){ // because we don't need to go till last we need second last
            if(arr[j] > arr[j+1]){
                let temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
                isSwap = true
            }
            if(isSwap === false) return;
        }
    }
}

let arr = [5,4,3,2,1]
bubbleSort(arr);
console.log(arr);

// time complexity for average case is => o(n) < average case <= o(n^2) which will take theta(n^2)
// for best case it is Ω(n) because of isSwap optimization we take

File: Sorting/Bubble sort/BubbleSort.java
public class BubbleSort{
    public static void bubbleSort(int[] arr){
        int n = arr.length;
        for(int i = 0; i < n; i++){
            boolean isSwap = false;
            for(int j = 0; j < n- i - 1; j++){
                // here n- i- 1 because we need to go till second last because we are comparing
                if(arr[j] > arr[j+1]){
                    int temp = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                    isSwap = true;
                } 
            }
            if(isSwap == false) return;
        }
    }

    public static void main(String[] args) {
        int[] arr = {10,3,-2,-1,4,3,2,1};
        bubbleSort(arr);

        for(int i: arr){
            System.out.println(i);
        }
    }
}